Allow executioners to enable exodus output

recuero opened the pull request:
Closes #25009 

@lynnmunday

recuero committed:
Allow per-iteration output in SteadyAndAdjoint solve from optimization solver (#25009)

This is a mildly invasive way of forcing proper Exodus iteration output in
steady and adjoint solves. We are casting in the Exodus object and passing
iteration number information back and forth from main app to child app.

There is a test that exemplifies the usage of this feature.

Restricts most changes to optimization module.

recuero committed:
Add documentation

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1660472/) on 0b19993 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25011/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25011/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 1520115d59e75a5e50a5927239b74318a0975355`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1660526/) on bfd3085 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25011/site)

This comment will be updated on new commits.

recuero committed:
Add exodus object for stead and adjoint

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1661359/) on 9303cd1 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>d2ef20</th>
      <th colspan=4>#25011 9303cd</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.59%</td>
      <td>85.59%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>89529</td>
      <td>89533</td>
      <td>+4</td>
      <td>4</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15073</td>
      <td>15072</td>
      <td>-1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25011/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25011/coverage/framework)


## Modules coverage

### Optimization

<table>
  <thead>
    <tr>
      <th></th>
      <th>d2ef20</th>
      <th colspan=4>#25011 9303cd</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.58%</td>
      <td>89.35%</td>
      <td>-0.23%</td>
      <td>81.82%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>1316</td>
      <td>1351</td>
      <td>+35</td>
      <td>36</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>153</td>
      <td>161</td>
      <td>+8</td>
      <td>8</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25011/coverage_diff/optimization)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25011/coverage/optimization)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25011/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25011/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25011/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25011/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25011/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25011/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25011/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25011/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25011/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25011/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/25011/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25011/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25011/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25011/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25011/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25011/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25011/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25011/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25011/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25011/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25011/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25011/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25011/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25011/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25011/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25011/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25011/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25011/coverage/xfem)

</details>

## Warnings

- `optimization` new line coverage rate 81.82% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud commented:
what are the desired execute_on flags for your output?
why does the boolean for "exodus_file_changed" not trigger for incrementing the file number?

recuero commented:
The flag that outputs the desired information for the `SteadyAndAdjoint` executioner is the `TIMESTEP_END`. 

`_exodus_mesh_changed` appears to refer to a change in the mesh, e.g. adaptivity. We would like to output the various iterations results within the same mesh in the subapp. If you just use `Exodus` to output results, they get overwritten (i.e. `_exodus_mesh_changed = false`).

GiudGiud commented:
What if you triggered the creation of a new file?
That boolean could be renamed to allow for these circumstances to create a new file

recuero commented:
Triggering the creation of a new file would require an additional input parameter to set `_exodus_mesh_changed` to `true`?

recuero commented:
The iteration number is also necessary to output a different time. The time step is not enough to distinguish between `SteadyAndAdjoint` outputs.

GiudGiud commented:
An input parameter or some other mechanic to allow the switch between the steady and transient. 

The timestep is not enough within a single file yes. But ought to be enough with two files

recuero commented:
> An input parameter or some other mechanic to allow the switch between the steady and transient.

Other mechanics could be figuring out the situation we are in by casting the executioner type (i.e. `SteadyAndAdjoint` in this case). However, the executioner we are interested in lives in the optimization module. Which, I believe, makes it design-wise less desirable.

> The timestep is not enough within a single file yes. But ought to be enough with two files

If we increment `_file_num` but keep the same time step, then the output cannot be visualized in Paraview.

If you prefer that I make the changes in the `Exodus` object, I can do that. The reason why I went the route of creating a particular type of exodus output in the optimization module is that I think it will provide more flexibility without adding API purely specific to optimization in the framework.

GiudGiud commented:
> If we increment _file_num but keep the same time step, then the output cannot be visualized in Paraview.

why is that? we dont get a new file?

> Other mechanics could be figuring out the situation we are in by casting the executioner type (i.e. SteadyAndAdjoint in this case). However, the executioner we are interested in lives in the optimization module. Which, I believe, makes it design-wise less desirable.

yes casting the executioner to select from a list is not desirable, especially outside the optimization module
But maybe executioners could trigger a new file on outputs for example. When the SteadyAndAdjoint does the transition from one to the other for example

recuero commented:
> why is that? we dont get a new file?

We would indeed get a new file. But Paraview seems to rely on a time step value to read successive files. That time step value would be an iteration number that we can only get from optimization type of solves.

> But maybe executioners could trigger a new file on outputs for example. When the SteadyAndAdjoint does the transition from one to the other for example

Yes, that's a possibility. To implement that, we'd also need to account for the iteration number (i.e. my answer right above). Which could be done by adding optimization module-specific API in the framework.

lindsayad commented:
We support outputting Exodus files on nonlinear iterations, so fixed point would make sense as well

lindsayad commented:
I need to find the code for how we do the nonlinear-iteration file naming. That could serve as an example or we could consolidate how we do that into a new "customStepNumber" routine

lindsayad commented:
I'm also curious about @zachmprince's thoughts as he was the original developer of this executioner and we've talked about this Exodus limitation

recuero commented:
I moved all the output logic to `Exodus`. I can improve/expand on this with your suggestions. I can probably add a framework test with whatever use case makes sense (i.e. fixed point iteration or others).

loganharbour commented:
@dschwen is actively working on this

recuero commented:
I believe this is what @loganharbour is referring to: https://github.com/idaholab/moose/pull/24706. This PR is intended to tackle a different sort of capability enhancement.

dschwen commented:
I disagree with the direction this PR took. A capability that is designed to work specifically with the optimization solver is presented as general. The PR introduces a second scheme for additional outputs besides the execute on flags. This should just be a derived Exodus output object in the optimization module.

recuero committed:
Reduce code duplication in child class

recuero commented:
I have switched back to the original design proposal. For the immediate needs, inheriting from `Exodus` in the optimization module and overriding a couple of parameters does the job with minimal disruption in the `Exodus` object. 

I can iterate more as we reach a consensus.

moosebuild commented:
Job [Min clang](https://civet.inl.gov/job/1665657/) on 52a79c5 : invalidated by @recuero

GiudGiud commented:
Doubt we ll reach consensus in this direction. Let s work on convincing Daniel he s wrong

dschwen commented:
The original design Antonio reverted to provides an API that can be augmented in the future to deal with fixed point iteration output. No need to cram that into this PR. As is the PR separates module and framework responsibilities cleanly. The `ExodusSteadyAndAdjoint` requires a specific executioner from the optimization module, and is itself in that module (the output should actually error, if the current executioner is not a match).

recuero committed:
ExodusSteadyAndAdjoint only works with SteadyAndAdjoint executioner.

Addresses Daniel's review.

The ExodusSteadyAndAdjoint output only produces a per optimization
itearation type of output. If used with a different executioner,
it will errour out.

lindsayad commented on framework/include/outputs/Exodus.h:
```diff
@@ -180,12 +190,14 @@ class Exodus : public OversampleOutput
   /// Storage for input file record; this is written to the file only after it has been initialized
   std::vector<std::string> _input_record;
 
+protected:
```
I have been guilty of this myself in the past, but let's try really hard to avoid interweaving access levels. Can we move the protected items ahead in the initializer list?

recuero replied:
Should be done now.

lindsayad commented on modules/optimization/src/executioners/OptimizeSolve.C:
```diff
@@ -251,6 +253,17 @@ OptimizeSolve::setTaoSolutionStatus(double f, int its, double gnorm, double cnor
+  // Pass down the iteration number if the subapp is of the SteadyAndAdjoint type.
+  // This enables exodus per-iteration output.
+  for (auto & sub_app : _app.getExecutioner()->feProblem().getMultiAppWarehouse().getObjects())
+  {
+    if (dynamic_cast<SteadyAndAdjoint *>(sub_app->getExecutioner(0)))
+      dynamic_cast<SteadyAndAdjoint *>(sub_app->getExecutioner(0))
+          ->setIterationNumberOutput(_optimize_iteration_number);
+  }
```
```suggestion
    if (auto steady_and_adjoint = dynamic_cast<SteadyAndAdjoint *>(sub_app->getExecutioner(0)))
     steady_and_adjoint->setIterationNumberOutput(_optimize_iteration_number);
```

lindsayad commented on modules/optimization/src/outputs/ExodusSteadyAndAdjoint.C:
```diff
@@ -0,0 +1,43 @@
+Real
+ExodusSteadyAndAdjoint::getTimeStepForOutput()
+{
+  return _steady_and_adjoint_exec ? _steady_and_adjoint_exec->getIterationNumberOutput() : time();
```
```suggestion
  return _steady_and_adjoint_exec->getIterationNumberOutput();
```
since you are erroring above

lindsayad commented on modules/optimization/test/tests/optimizationreporter/point_loads/forward_and_adjoint_iteration_output.i:
```diff
@@ -0,0 +1,132 @@
+
+[Outputs]
+  console = false
+
```
```suggestion
```

lindsayad commented on modules/optimization/test/tests/optimizationreporter/point_loads/tests:
```diff
@@ -43,5 +43,18 @@
+      exodiff = 'main_auto_adjoint_out_forward0_exodus.e-s005'
+      # Optimize executioner does not support multiple threads
+      max_threads = 1
+      rel_err = 0.1
+      abs_zero = 10
```
why are these tolerances so high?

recuero replied:
These are the same tolerances used for the original setup. I think they can be tightened, it passes locally with default ones. Hopefully it also does on CIVET.

lindsayad commented on modules/optimization/include/executioners/OptimizeSolve.h:
```diff
@@ -149,6 +149,9 @@ class OptimizeSolve : public SolveObject
   /// Parameters (solution)
   std::unique_ptr<libMesh::PetscVector<Number>> _parameters;
 
+  /// Iteration number for the forward problem
```
This is really for the forward problem/solve?

recuero replied:
It's for the optimization process; for steady and adjoint should be forward-adjoint solve. But Lynn wanted to extend it to plain `Steady` too. So I switched the comment to "Optimization iteration number".

lynnmunday commented:
@recuero Would it be hard to add an exodus output of just a regular steady executioner for each optimization iteration?  Sometimes we just use a steady executioner for the forward problem instead of the SteadyAndAdjoint executioner.

recuero committed:
Address Alex's comment

recuero committed:
Extend per-iteration output to plain Steady solves.

This commit addresses Lynn's request to have this capability for
Steady solves as well. As of this commit, the Exodus object that
enables the per-iteration output is ExodusOptimizationSteady, which
would work for Steady and SteadyAndAdjoint solves.

The ability for the user to try to output per-iteration is restricted
to the optimization module.

lindsayad commented:
@GiudGiud I think this is a good PR and should go in. If we come up with a design improvement later then we can refactor. I don't think there's anything here that would limit future flexibility

GiudGiud commented on framework/include/executioners/Steady.h:
```diff
@@ -45,6 +45,15 @@ class Steady : public Executioner
   virtual bool lastSolveConverged() const override { return _last_solve_converged; }
 
+  /**
+   * Get optimization iteration number from main app (when Steady solves an optimization problem)
```
```suggestion
   * Get a general iteration number for the purpose of outputting from the main app, useful in the presence of a nested solve
```

GiudGiud commented on framework/include/executioners/Steady.h:
```diff
@@ -45,6 +45,15 @@ class Steady : public Executioner
+   */
+  virtual unsigned int getIterationNumberOutput() const;
+  /**
+   * Set optimization iteration number from main app (when Steady solves an optimization problem)
```
same

GiudGiud replied:
mention the subapp as the current use case

lindsayad replied:
These APIs are probably what I like least about the current PR as it exists only for the purpose of the Optimization module ... no framework Steady cases have multiple output iterations (with the exception of per-solve-iterations). Perhaps the exception is the best reason for the existence of these APIs

recuero replied:
I agree with Alex. @GiudGiud mentioned fixed point iteration solves as a use case for which we may want to have additional, tailored output: Would it make sense to rename that API as something like (get/set)OutputTimeStep and move it up the executioner hierarchy? That may be good if it's going to find more use.

Alternatively, if multiple usage isn't going to happen for that executioner API, we could create a `Steady` executioner class in the optimization module that encapsulates the behavior that the optimization module expects. I think this may not be a great idea since Guillaume thought what we are trying to do here can be extended to other cases.

GiudGiud commented on framework/include/executioners/Steady.h:
```diff
@@ -54,6 +63,9 @@ class Steady : public Executioner
   Real & _time;
 
+  /// Iteration number obtained from the main application
+  unsigned int _optimization_iteration_number;
```
```suggestion
  unsigned int _output_iteration_number;
```

GiudGiud commented on framework/include/executioners/Steady.h:
```diff
@@ -45,6 +45,15 @@ class Steady : public Executioner
+  /**
+   * Get optimization iteration number from main app (when Steady solves an optimization problem)
+   */
+  virtual unsigned int getIterationNumberOutput() const;
```
they are one-liners they can be defined here in the header

GiudGiud commented on framework/src/executioners/Steady.C:
```diff
@@ -126,3 +127,15 @@ Steady::checkIntegrity()
+unsigned int
+Steady::getIterationNumberOutput() const
+{
+  return _optimization_iteration_number;
```
move to header if one-line

GiudGiud commented on modules/optimization/include/executioners/OptimizeSolve.h:
```diff
@@ -149,6 +149,9 @@ class OptimizeSolve : public SolveObject
   /// Parameters (solution)
   std::unique_ptr<libMesh::PetscVector<Number>> _parameters;
 
+  /// Optimization teration number
+  unsigned int _optimize_iteration_number;
```
why do we need 2 attributes for the same thing?

GiudGiud replied:
(in Steady you added the same)

GiudGiud commented on modules/optimization/src/outputs/ExodusOptimizationSteady.C:
```diff
@@ -0,0 +1,43 @@
+        "executioners to output a per-optimization iteration solution.");
+}
+
+void
+ExodusOptimizationSteady::incrementFileCounter()
+{
+  _file_num = _steady_exec->getIterationNumberOutput();
+}
+
+Real
+ExodusOptimizationSteady::getTimeStepForOutput()
+{
+  return _steady_exec->getIterationNumberOutput();
+}
```
this could just be the behavior of the Exodus output and you can delete that class

GiudGiud replied:
in fact if someone overrode these methods in their executioner, they could be surprised not to see the right behavior in their Exodus output

GiudGiud replied:
@dschwen @lindsayad this is where we need to decide if these two lines (that likely should be in the base class as well, since they rely on something a dev would have implemented in their executioner) are worth creating a new class for

I dont like naming outputs after Executioners.

GiudGiud replied:
I guess to move this up to "Exodus" we would need to move up the "output step number" APIs to Executioner, as we don't really want to condition the call to the step number routine in Exodus to the type of the executioner. 

So I'm fine with this object for now.
I think the solution that keeps in line with our design principles would be to either rework the SetupInterface (execute_on) or create a new interface between Outputs and Executioner/Executors that handles the determination of the step number.

lindsayad replied:
The thing is that the majority of the time, except when a user wants to do per-solve-iteration debugging and in this optimization module use case, the exodus output file name has nothing to do with executioner data. In the majority of cases, the exodus output file name only depends on adaptivity data, which we divorced from the executioner block deliberately (although that syntax still exists). So if we wanted to tie Exodus output (or generically any output) to executioner data, we may need to make the executioner aware of more things, or the executioner will have to go through a chain of calls to get the adaptivity step, etc.

GiudGiud commented on modules/optimization/test/tests/optimizationreporter/material/adjoint_iteration_output.i:
```diff
@@ -0,0 +1,112 @@
+[Mesh]
```
keep the test here for now until the design is set

but if we stay with this "ExodusOptimizationSteady" object they will need to be moved to a 
optizimation/test/tests/outputs/exodus_optimization_steady folder.

GiudGiud commented:
This has removed the code duplication of the Exodus class so I dont think this is the original design at all.

I think it's in good shape, the APIs introduced to get an "output step number" from a Steady executioner make sense to me.

YaqiWang commented on framework/src/outputs/Exodus.C:
```diff
@@ -531,3 +537,9 @@ Exodus::handleExodusIOMeshRenumbering()
 }
+
+Real
+Exodus::getTimeStepForOutput()
```
Maybe rename this with `getOutputTime`? Time step to me means an integer step number. This function represents a right direction in my opinion and should be used by all outputs. So I also suggest to move it to the base class and let all outputs stop calling `time()` directly.

YaqiWang commented on framework/src/executioners/Steady.C:
```diff
@@ -126,3 +127,15 @@ Steady::checkIntegrity()
+}
+
+void
+Steady::setIterationNumberOutput(unsigned int iteration_number)
```
This two functions look to me belonging to output system, either in Outputs or OutputWarehouse.

recuero replied:
If I understand well, the executioner would update an output system member variable that can be used without having the output derived objects pulled the executioner. That makes sense to me. I'd rather not make this change now in this PR. I suggest we leave that for when we further exercise this "special" output capability.

YaqiWang replied:
This interface function seems the hot spot from reviewers ;-) This really looks like a short term solution. If we can have a little bit more visionary design here, it would be better. I do not have super strong opinions on this.

recuero committed:
Address Guillaume's review

recuero committed:
Address Yaqi's review

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1670196/) on 441fa3c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25011/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25011/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 98c121565021040d042a982ab2ca0726c512b7b2`

recuero committed:
Remove outdated file from optimization_reporter folder.

And update docs with new test location.

lindsayad commented:
@GiudGiud I am waiting on your re-review for moving this forward. I know you are on vacation, but I also don't know what the timelines are for those who want to use this PR

GiudGiud commented:
on it now!

GiudGiud commented:
why do we have both?
```
void
ExodusOptimizationSteady::incrementFileCounter()
{
  _file_num = _steady_exec->getIterationNumberOutput();
}

Real
ExodusOptimizationSteady::getOutputTime()
{
  return _steady_exec->getIterationNumberOutput();
}
```

do we want multiple opt steps in a file or multiple files with each opt step? or both??

GiudGiud commented on modules/optimization/test/tests/outputs/exodus_optimization_steady/gold/adjoint.e-s007:
this file has an adjoint that s identically 0, is that normal?

lynnmunday replied:
@recuero could you use adjoint.e-s002 instead of 007, it should have a larger adjoint variable output to the exodus file.  When the optimization is converged, the misfit between the simulation and measurement data will get really small, making the adjoint variable really small.

recuero replied:
The entire file is being tested now. 

Guillaume's observation was correct. The output was null because the `execute_on` was not appropriate. I am now defaulting to `TIMESTEP_END` in the object itself.

recuero committed:
Exodus output of optimization iterations takes place in one single file.

Also, make sure output takes place at time step end for proper output.

This commit addresses Guillaume's re-review and discussions with Lynn.

recuero commented:
@lynnmunday was okay with having single output file iteration output, so switched to that strategy.

GiudGiud reviewed (approved):
fine with me once it passes tests.

these may not be the right APIs on the executioner, I dont know. I feel like we need something like this when we will have complex executor-combinations

moosebuild commented:
Job [Conda build (ARM Mac)](https://civet.inl.gov/job/1695576/) on d3bd395 : invalidated by @lindsayad

lindsayad commented:
@recuero feel free to merge when you're ready
